### PR TITLE
Component for creating DPM and Allowance flow

### PR DESCRIPTION
**Pull request by marcinciarka**
# Component + hook for creating DPM and Allowance flow

  
## Changes 👷‍♀️
- added new hook - `useFlowState`. The purpose of this hook is to provide access to internals of handling flow states: creating DPM proxy and allowance.
- added new component `FlowSideBar` which can consume `flowState` from the hook above. This provides UI needed for proxy and allowance states.

## How to use it? 🔧 
Typical use case:
```
  const flowState = useFlowState({
    amount: yourDataSource.amount,
    token: yourDataSource.collateral,
    onEverythingReady(params) {
      // we have everything we need, pick the first
      // DPM proxy from "availableProxies" and proceed with your flow
      console.log('onEverythingReady', params)
    },
  })
```
```
  <FlowSidebar {...flowState} />
```
If you don't want to use `onEverythingReady` for whatever reason you can react to the hook changes:
```
{
    dpmMachine, // internals
    allowanceMachine, // internals
    availableProxies, // list of available proxies ⭐
    walletAddress, // just for convenience
    amount, // just for convenience
    token, // just for convenience
    isProxyReady, // self explanatory
    isWalletConnected, // self explanatory
    isAllowanceReady, // self explanatory
    isLoading, // just for the allowance loading state
    isEverythingReady, // self explanatory
  }
```
⭐ internals use first (`[0]`) for allowance and so should you for your flow

For manage flows - everything stays the same, just provide `existingProxy` to the hook and it will be available in `availableProxies`, the hook will ignore getting the list of proxies. This is helpful for the allowance flow (which might need to be updated if user picked minimum allowance for example).
  

  
## How to test 🧪
- use it, tell me what you think and how you feel 💡

### CODE REVIEW

File: helpers/useFlowState.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    proxyConsumed$,
+    allowanceForAccount$,
+  } = useAppContext()
+  const { context$ } = useAppContext()
+  const { stateMachine: dpmMachine } = setupDpmContext(dpmAccountStateMachine)
+  const { stateMachine: allowanceMachine } = setupAllowanceContext(allowanceStateMachine)
+
+  // wallet connection + DPM proxy machine
+  useEffect(() => {
+    const walletConnectionSubscription = context$.subscribe(({ status, account }) => {
```
Comment (jakubswierczek) [inline helpers/useFlowState.ts:56]: `useObservable`

File: helpers/useFlowState.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    walletAddress: UserFlowStateReturnType['walletAddress']
+    amount: UserFlowStateReturnType['amount']
+    token: UserFlowStateReturnType['token']
+    isProxyReady: UserFlowStateReturnType['isProxyReady']
+    isWalletConnected: UserFlowStateReturnType['isWalletConnected']
+    isAllowanceReady: UserFlowStateReturnType['isAllowanceReady']
+  }) => void
+}
+
+export function useFlowState({
```
Comment (jakubswierczek) [inline helpers/useFlowState.ts:29]: why do we have two responsibilities in one flow? 
we should have one flow for DPMs and one for allowance.

File: helpers/useFlowState.ts [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+  amount,
+  token,
+  onEverythingReady,
+  existingProxy,
+}: UseFlowStateProps) {
```
Comment (jakubswierczek) [inline helpers/useFlowState.ts:34]: I think we should take `owner` as a parameter. Not getting wallet address from `context$`.

File: helpers/useFlowState.ts [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+import { setupAllowanceContext, setupDpmContext } from './dummyStateMachine'
+import { zero } from './zero'
+
+type UserFlowStateReturnType = ReturnType<typeof useFlowState>
+
+type UseFlowStateProps = {
+  amount?: BigNumber
+  token?: string
+  existingProxy?: string
+  onEverythingReady?: (params: {
```
Comment (jakubswierczek) [inline helpers/useFlowState.ts:18]: please make a type for an event

File: helpers/useFlowState.ts [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    allowanceMachine,
+    availableProxies,
+    walletAddress,
+    amount,
+    token,
+    isProxyReady: !!availableProxies.length,
+    isWalletConnected,
+    isAllowanceReady,
+    isLoading, // just for the allowance loading state
+    isEverythingReady: isAllowanceReady, // just for convenience, allowance is always the last step
```
Comment (jakubswierczek) [inline helpers/useFlowState.ts:188]: What about the situation when the user has approval for a given amount?

**Review by piotrkonowrocki — CHANGES_REQUESTED** [review]
Few comments about allowance stage:
- Before I select any option, the button should be disabled.
- Loader never stops spinning. Also, there's a funny effect, every time when I hover over anything inside of a sidebar it resets and try spinning from the top 💃
image

After selecting amount and confirming transaction, the message in sidebar is incomplete. Shouldn't it say "(...) Your Smart Defi Account will be able to interact with up to: **AMOUNT**", since we know what was selected?
image
image

Shouldn't there be something else on allowance success screen? An image similar to the on on dpm success screen?
image

When I click continue on success screen, after a second or so success screen is shown up again. Data from `onEverthingReady` seems to be correct:
image

**Review by jakubswierczek — CHANGES_REQUESTED** [review]

File: components/FlowSidebar.tsx [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+  const allowanceConsidered = token !== 'ETH' && amount
+
+  if (!isWalletConnected) {
+    return <NoConnectionStateView noConnectionContent={noConnectionContent} />
+  }
+  if (!isProxyReady && !isAllowanceReady) {
+    switch (true) {
+      case dpmState.matches('idle'):
+      case dpmState.matches('txFailure'):
+        return <DPMInfoStateView state={dpmState} send={dpmSend} />
```
Comment (piotrkonowrocki) [inline components/FlowSidebar.tsx:226]: Allowance is nicely split and separated into `AllowanceView`, I suggest moving DPM related components from this file to something similar.

File: helpers/useFlowState.ts (hunk 2 above)
Comment (piotrkonowrocki) [reply helpers/useFlowState.ts:29]: This is expected behavior. For our current needs there isn't a situation where you need only DPM or only allowance flows - they always come as a pair.

We also spoke that in the future that could be refined into a wrapped that merges two separated flow, but for now this is exactly what we need.

File: helpers/useFlowState.ts [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    isProxyReady: UserFlowStateReturnType['isProxyReady']
+    isWalletConnected: UserFlowStateReturnType['isWalletConnected']
+    isAllowanceReady: UserFlowStateReturnType['isAllowanceReady']
+  }) => void
+}
+
+export function useFlowState({
+  amount,
+  token,
+  onEverythingReady,
```
Comment (piotrkonowrocki) [inline helpers/useFlowState.ts:32]: We are going to need one more callback: `onBack`. If function is passed, there should be a `textButton` passed to sidebar that would trigger this callback. Thanks to that you can always get back from allowance or dpm step back to from wherever you came from.

And I guess that button should be visible only on first steps of both flows. If we put it on progress, success or error screen then it doesn't make any sense.

**Review by piotrkonowrocki — CHANGES_REQUESTED** [review]

File: helpers/useFlowState.ts (hunk 3 above)
Comment (marcinciarka) [reply helpers/useFlowState.ts:34]: We're never gonna check the DPM accounts of another address than wallet and this simplifies things. If you have a strong opinion please reopen :v:

File: helpers/dummyStateMachine.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+      }
+    })
+
+  const service = useInterpret(machine, { parent: parentService }).start()
+
+  const states$ = from(service)
+  return {
+    stateMachine: service,
+    states$,
+    dpmAccounts$: createdDpmAccount.asObservable(),
```
[the remark this answers is not in the document]
Comment (marcinciarka) [reply helpers/dummyStateMachine.ts:44]: Nope in both cases, removed.

File: helpers/useFlowState.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (marcinciarka) [reply helpers/useFlowState.ts:32]: This is ultimately _always_ the last step, thats why I named it like this.

File: helpers/useFlowState.ts (hunk 5 above)
Comment (marcinciarka) [reply helpers/useFlowState.ts:193]: There's a check for that in `allowanceSubscription`.

File: components/FlowSidebar.tsx (hunk 6 above)
Comment (marcinciarka) [reply components/FlowSidebar.tsx:226]: Done.

File: helpers/useFlowState.ts (hunk 7 above)
Comment (marcinciarka) [reply helpers/useFlowState.ts]: Done, `onGoBack`.

**Review by jakubswierczek — DISMISSED** [review]

File: helpers/useFlowState.ts (hunk 7 above)
Comment (piotrkonowrocki) [reply helpers/useFlowState.ts]: ❤️

**Review by piotrkonowrocki — APPROVED** [review]

**Pull request merged**